File: src/attestations/store.zig
// SPDX-License-Identifier: MIT OR Apache-2.0
//
// On-disk store for DSSE attestation envelopes.
// See docs/SPEC-ATTESTATIONS.md §3.
//
// Layout (rooted at the mkit directory the caller opens — typically
// `.mkit/`, not the `attestations/` subdir; every function opens
// subdirectories as needed):
//
//     <root-dir>/
//       attestations/
//         <64-hex-commit>/
//           <64-hex-att-id>.dsse
//
// Envelope bytes are written exactly as supplied. SPEC-ATTESTATIONS §3
// line 95 says the on-disk form terminates with a single `\n`, but that
// newline is the *encoder's* responsibility (the canonical JCS output
// has no trailing newline; the encoder in envelope.zig does not emit
// one either). We preserve byte-exact input so that
// `attestationId(bytes)` — the filename — and the file contents agree;
// appending a newline here would mean the file's blake3 no longer
// matches the att-id, breaking the content-addressing invariant.
//
// Durability pattern mirrors src/store.zig and
// src/transport/file.zig (commit 41be906, which introduced the
// containing-directory fsync to close the crash-durability gap after
// rename(2)):
//
//     createFileAtomic → writeAll → flush → file.sync → Atomic.replace
//     → fsync the commit directory (via libc fsync on dir.handle)
//
// `std.Io.Dir` in Zig 0.16 has no `sync` method, but `dir.handle` is
// the raw POSIX fd and `std.c.fsync` accepts directory descriptors on
// both Linux and Darwin. mkit links libc unconditionally so this is
// safe without further build changes.

const std = @import("std");
const Allocator = std.mem.Allocator;
const Io = std.Io;

const hash_mod = @import("../hash.zig");
const Hash = hash_mod.Hash;
const protocol = @import("../protocol.zig");
const envelope_mod = @import("envelope.zig");

/// Errors specific to the attestation store.
pub const Error = error{
    NotFound,
};

/// The subdirectory under `root_dir` that owns every envelope on disk.
pub const subdir = "attestations";

/// Envelope file extension. Keep in sync with SPEC-ATTESTATIONS §3.
pub const file_ext = ".dsse";

/// Safety cap on a single envelope read. A well-formed envelope with a
/// few signatures is well under a kilobyte; even an absurd multi-sig
/// case stays in the low tens of kilobytes. 1 MiB is a generous ceiling
/// that still protects against a corrupt or malicious pathname landing
/// a giant file in this tree and starving the reader.
pub const MAX_ENVELOPE_SIZE: usize = 1 * 1024 * 1024;

/// "<64-hex>" — the directory name used for a commit's attestations.
pub fn commitDirName(commit: Hash) [hash_mod.hex_len]u8 {
    return hash_mod.toHex(commit);
}

/// "<64-hex>.dsse" — the filename used for a single envelope.
pub fn attFileName(att_id: Hash) [hash_mod.hex_len + file_ext.len]u8 {
    var out: [hash_mod.hex_len + file_ext.len]u8 = undefined;
    const hex = hash_mod.toHex(att_id);
    @memcpy(out[0..hash_mod.hex_len], &hex);
    @memcpy(out[hash_mod.hex_len..], file_ext);
    return out;
}

/// Write a DSSE envelope for `commit`. Returns the att-id.
///
/// Idempotent via content-addressing: the filename is the BLAKE3 of
/// the bytes, so if the target already exists we return the same id
/// without rewriting.
pub fn writeAttestation(
    _: Allocator,
    io: Io,
    root_dir: Io.Dir,
    commit: Hash,
    envelope_bytes: []const u8,
) !Hash {
    if (envelope_bytes.len > MAX_ENVELOPE_SIZE) return error.ObjectTooLarge;

    const att_id = envelope_mod.attestationId(envelope_bytes);
    const commit_hex = commitDirName(commit);
    const file_name = attFileName(att_id);

    root_dir.createDir(io, subdir, .default_dir) catch |err| switch (err) {
        error.PathAlreadyExists => {},
        else => return err,
    };

    var att_root = try root_dir.openDir(io, subdir, .{});
    defer att_root.close(io);

    att_root.createDir(io, &commit_hex, .default_dir) catch |err| switch (err) {
        error.PathAlreadyExists => {},
        else => return err,
    };

    var commit_dir = try att_root.openDir(io, &commit_hex, .{});
    defer commit_dir.close(io);

    // Idempotency via content-addressing: the filename IS the BLAKE3 of
    // the bytes, so presence implies equality. One stat beats a full
    // read+compare.
    commit_dir.access(io, &file_name, .{}) catch |err| switch (err) {
        error.FileNotFound => {
            var atomic_file = try commit_dir.createFileAtomic(io, &file_name, .{ .replace = true });
            defer atomic_file.deinit(io);

            var buffer: [4096]u8 = undefined;
            var file_writer = atomic_file.file.writer(io, &buffer);
            try file_writer.interface.writeAll(envelope_bytes);
            try file_writer.interface.flush();
            try file_writer.file.sync(io);
            try atomic_file.replace(io);

            // fsync the containing directory so rename(2) survives a
            // power loss. Best-effort: tmpfs rejects dir-fd fsync.
            syncDir(commit_dir) catch {};
            return att_id;
        },
        else => return err,
    };
    return att_id;
}

/// Read the envelope for `commit` + `att_id`. Allocator-owned bytes.
pub fn readAttestation(
    allocator: Allocator,
    io: Io,
    root_dir: Io.Dir,
    commit: Hash,
    att_id: Hash,
) ![]u8 {
    const commit_hex = commitDirName(commit);
    const file_name = attFileName(att_id);

    var att_root = root_dir.openDir(io, subdir, .{}) catch |err| switch (err) {
        error.FileNotFound => return error.NotFound,
        else => return err,
    };
    defer att_root.close(io);

    var commit_dir = att_root.openDir(io, &commit_hex, .{}) catch |err| switch (err) {
        error.FileNotFound => return error.NotFound,
        else => return err,
    };
    defer commit_dir.close(io);

    return commit_dir.readFileAlloc(io, &file_name, allocator, .limited(MAX_ENVELOPE_SIZE)) catch |err| switch (err) {
        error.FileNotFound => return error.NotFound,
        else => return err,
    };
}

/// List every att-id attached to `commit`, sorted ascending
/// byte-wise on the raw 32-byte hash. An unattested commit (no
/// directory) returns an empty slice — NOT an error, per
/// SPEC-ATTESTATIONS §7.1 (a commit with zero and a commit with five
/// attestations are indistinguishable at the commit layer).
pub fn listAttestations(
    allocator: Allocator,
    io: Io,
    root_dir: Io.Dir,
    commit: Hash,
) ![]Hash {
    const commit_hex = commitDirName(commit);

    var att_root = root_dir.openDir(io, subdir, .{}) catch |err| switch (err) {
        error.FileNotFound => return &[_]Hash{},
        else => return err,
    };
    defer att_root.close(io);

    var commit_dir = att_root.openDir(io, &commit_hex, .{ .iterate = true }) catch |err| switch (err) {
        error.FileNotFound => return &[_]Hash{},
        else => return err,
    };
    defer commit_dir.close(io);

    var ids: std.ArrayList(Hash) = .empty;
    errdefer ids.deinit(allocator);

    var iter = commit_dir.iterate();
    while (try iter.next(io)) |entry| {
        if (entry.kind != .file) continue;
        const id = parseAttFileName(entry.name) orelse continue;
        try ids.append(allocator, id);
    }

    const slice = try ids.toOwnedSlice(allocator);
    std.sort.pdq(Hash, slice, {}, protocol.hashLessThan);
    return slice;
}

/// Remove a single attestation. Idempotent: removing a non-existent
/// att-id (or a commit that has no attestations at all) returns
/// successfully. When the commit directory becomes empty, it is
/// removed as well — tolerating `DirNotEmpty` if a concurrent writer
/// landed a new envelope in the interim.
pub fn removeAttestation(
    io: Io,
    root_dir: Io.Dir,
    commit: Hash,
    att_id: Hash,
) !void {
    const commit_hex = commitDirName(commit);
    const file_name = attFileName(att_id);

    var att_root = root_dir.openDir(io, subdir, .{}) catch |err| switch (err) {
        error.FileNotFound => return,
        else => return err,
    };
    defer att_root.close(io);

    {
        var commit_dir = att_root.openDir(io, &commit_hex, .{}) catch |err| switch (err) {
            error.FileNotFound => return,
            else => return err,
        };
        defer commit_dir.close(io);

        commit_dir.deleteFile(io, &file_name) catch |err| switch (err) {
            error.FileNotFound => {},
            else => return err,
        };
    }

    // Attempt to garbage-collect the now-possibly-empty commit dir.
    att_root.deleteDir(io, &commit_hex) catch |err| switch (err) {
        error.DirNotEmpty, error.FileNotFound => {},
        else => return err,
    };
}

// -----------------------------------------------------------------------------
// Internals
// -----------------------------------------------------------------------------

const parseAttFileName = protocol.parseAttestationFilename;

/// fsync a directory file descriptor. See the module doc comment for
/// why this is needed and why it uses libc directly.
fn syncDir(dir: Io.Dir) !void {
    if (std.c.fsync(dir.handle) != 0) return error.SyncFailed;
}

// -----------------------------------------------------------------------------
// Tests
// -----------------------------------------------------------------------------

const testing = std.testing;

/// Build a fake `Hash` from a seed byte, so tests don't depend on any
/// particular real commit hash.
fn fakeHash(seed: u8) Hash {
    var h: Hash = undefined;
    for (&h, 0..) |*b, i| b.* = seed +% @as(u8, @intCast(i));
    return h;
}

test "write/read roundtrip" {
    const allocator = testing.allocator;
    const io = testing.io;

    var tmp = testing.tmpDir(.{});
    defer tmp.cleanup();

    const commit = fakeHash(0x11);
    const envelope = "x" ** 50;

    const att_id = try writeAttestation(allocator, io, tmp.dir, commit, envelope);
    try testing.expectEqual(envelope_mod.attestationId(envelope), att_id);

    const got = try readAttestation(allocator, io, tmp.dir, commit, att_id);
    defer allocator.free(got);
    try testing.expectEqualSlices(u8, envelope, got);
}

test "write is idempotent: same bytes produce same id and one file" {
    const allocator = testing.allocator;
    const io = testing.io;

    var tmp = testing.tmpDir(.{});
    defer tmp.cleanup();

    const commit = fakeHash(0x22);
    const envelope = "hello envelope";

    const a = try writeAttestation(allocator, io, tmp.dir, commit, envelope);
    const b = try writeAttestation(allocator, io, tmp.dir, commit, envelope);
    try testing.expectEqual(a, b);

    const ids = try listAttestations(allocator, io, tmp.dir, commit);
    defer allocator.free(ids);
    try testing.expectEqual(@as(usize, 1), ids.len);
    try testing.expectEqual(a, ids[0]);
}

test "list returns att-ids sorted ascending" {
    const allocator = testing.allocator;
    const io = testing.io;

    var tmp = testing.tmpDir(.{});
    defer tmp.cleanup();

    const commit = fakeHash(0x33);

    var envelopes: [5][16]u8 = undefined;
    var written: [5]Hash = undefined;
    for (&envelopes, 0..) |*e, i| {
        @memcpy(e, "envelope-00-tail");
        e.*[9] = '0' + @as(u8, @intCast(i));
        written[i] = try writeAttestation(allocator, io, tmp.dir, commit, e);
    }

    const ids = try listAttestations(allocator, io, tmp.dir, commit);
    defer allocator.free(ids);
    try testing.expectEqual(@as(usize, 5), ids.len);

    // Sorted ascending byte-wise.
    var j: usize = 1;
    while (j < ids.len) : (j += 1) {
        try testing.expect(std.mem.order(u8, &ids[j - 1], &ids[j]) == .lt);
    }

    // Every written id must appear in the result set.
    for (written) |w| {
        var found = false;
        for (ids) |id| {
            if (std.mem.eql(u8, &w, &id)) {
                found = true;
                break;
            }
        }
        try testing.expect(found);
    }
}

test "list on unattested commit returns empty slice, no error" {
    const allocator = testing.allocator;
    const io = testing.io;

    var tmp = testing.tmpDir(.{});
    defer tmp.cleanup();

    const commit = fakeHash(0x44);
    const ids = try listAttestations(allocator, io, tmp.dir, commit);
    defer allocator.free(ids);
    try testing.expectEqual(@as(usize, 0), ids.len);

    // Same result even after another commit's attestations exist.
    const other = fakeHash(0x55);
    _ = try writeAttestation(allocator, io, tmp.dir, other, "some envelope");

    const ids2 = try listAttestations(allocator, io, tmp.dir, commit);
    defer allocator.free(ids2);
    try testing.expectEqual(@as(usize, 0), ids2.len);
}

test "remove is idempotent on missing att-id" {
    const io = testing.io;
    var tmp = testing.tmpDir(.{});
    defer tmp.cleanup();

    const commit = fakeHash(0x66);
    const missing = fakeHash(0x77);

    // No attestations dir at all.
    try removeAttestation(io, tmp.dir, commit, missing);

    // attestations dir exists, commit dir doesn't.
    const allocator = testing.allocator;
    const other = fakeHash(0x88);
    _ = try writeAttestation(allocator, io, tmp.dir, other, "keep me");
    try removeAttestation(io, tmp.dir, commit, missing);

    // commit dir exists but the specific att-id doesn't.
    _ = try writeAttestation(allocator, io, tmp.dir, commit, "present");
    try removeAttestation(io, tmp.dir, commit, missing);

    // Nothing we had should have been clobbered.
    const ids = try listAttestations(allocator, io, tmp.dir, commit);
    defer allocator.free(ids);
    try testing.expectEqual(@as(usize, 1), ids.len);
}

test "remove cleans up an empty commit dir" {
    const allocator = testing.allocator;
    const io = testing.io;
    var tmp = testing.tmpDir(.{});
    defer tmp.cleanup();

    const commit = fakeHash(0x99);
    const att_id = try writeAttestation(allocator, io, tmp.dir, commit, "only one");

    try removeAttestation(io, tmp.dir, commit, att_id);

    // list on the now-gone commit dir should be empty and not error.
    const ids = try listAttestations(allocator, io, tmp.dir, commit);
    defer allocator.free(ids);
    try testing.expectEqual(@as(usize, 0), ids.len);

    // The commit dir itself is gone.
    var att_root = try tmp.dir.openDir(io, subdir, .{});
    defer att_root.close(io);
    const commit_hex = commitDirName(commit);
    try testing.expectError(error.FileNotFound, att_root.openDir(io, &commit_hex, .{}));
}

test "list ignores non-.dsse and non-hex-stem files" {
    const allocator = testing.allocator;
    const io = testing.io;

    var tmp = testing.tmpDir(.{});
    defer tmp.cleanup();

    const commit = fakeHash(0xAA);
    const good_id = try writeAttestation(allocator, io, tmp.dir, commit, "legit envelope");

    // Drop two junk files directly into the commit dir.
    {
        var att_root = try tmp.dir.openDir(io, subdir, .{});
        defer att_root.close(io);
        const commit_hex = commitDirName(commit);
        var commit_dir = try att_root.openDir(io, &commit_hex, .{});
        defer commit_dir.close(io);

        // (a) Wrong extension entirely.
        try commit_dir.writeFile(io, .{ .sub_path = "notes.txt", .data = "ignored" });
        // (b) Right extension, wrong-shaped stem (too short + non-hex char).
        try commit_dir.writeFile(io, .{ .sub_path = "zzz.dsse", .data = "ignored" });
        // (c) Right length, right extension, but non-hex stem ('g' is invalid).
        const bad_hex_stem = "g" ++ ("0" ** 63) ++ ".dsse";
        try commit_dir.writeFile(io, .{ .sub_path = bad_hex_stem, .data = "ignored" });
    }

    const ids = try listAttestations(allocator, io, tmp.dir, commit);
    defer allocator.free(ids);
    try testing.expectEqual(@as(usize, 1), ids.len);
    try testing.expectEqual(good_id, ids[0]);
}

test "attFileName and commitDirName are deterministic hex" {
    const h = fakeHash(0xBB);
    const hex = hash_mod.toHex(h);

    const dir_name = commitDirName(h);
    try testing.expectEqualSlices(u8, &hex, &dir_name);

    const file_name = attFileName(h);
    try testing.expectEqualSlices(u8, &hex, file_name[0..64]);
    try testing.expectEqualStrings(".dsse", file_name[64..]);
}

File: rust/crates/mkit-attest/src/store.rs
//! On-disk store for DSSE attestation envelopes.
//!
//! See `docs/SPEC-ATTESTATIONS.md` §3. Layout (rooted at `<root>` —
//! typically `.mkit/`, NOT the `attestations/` subdir):
//!
//! ```text
//! <root>/
//!   attestations/
//!     <64-hex-commit>/
//!       <64-hex-att-id>.dsse
//! ```
//!
//! Envelope bytes are written exactly as supplied (no trailing newline
//! is appended). The att-id is `BLAKE3(envelope_bytes)`, which makes
//! the filename a stable, content-addressed function of the envelope —
//! re-writing identical bytes is a no-op (idempotent via stat).
//!
//! Durability mirrors `mkit-core::store`:
//! `tempfile::NamedTempFile::persist` (atomic rename) plus an
//! Unix-only `fsync` of the containing commit directory afterwards so
//! the dirent update survives a power loss.

use std::fs;
#[cfg(unix)]
use std::fs::File;
use std::io::Write;
use std::path::{Path, PathBuf};
use std::process;
use std::sync::atomic::{AtomicU64, Ordering};

use tempfile::NamedTempFile;

use crate::Error;
use crate::envelope as env_mod;
use mkit_core::Hash;
use mkit_core::hash::{self as hash_mod, HEX_LEN};

/// Directory under `<root>` that owns every envelope on disk.
pub const SUBDIR: &str = "attestations";

/// Envelope file extension. Keep in sync with SPEC-ATTESTATIONS §3.
pub const FILE_EXT: &str = ".dsse";

/// Safety cap on a single envelope read. A well-formed envelope with a
/// few signatures is well under a kilobyte; even an absurd multi-sig
/// case stays in the low tens of kilobytes. 1 MiB is a generous ceiling.
pub const MAX_ENVELOPE_SIZE: usize = 1024 * 1024;

/// Build the path for a given commit + att-id under `root`.
///
/// `root` is the `.mkit/` directory (the `attestations/` subdir is
/// appended internally).
#[must_use]
pub fn envelope_path(root: &Path, commit: &Hash, att_id: &Hash) -> PathBuf {
    root.join(SUBDIR)
        .join(hash_mod::to_hex(commit))
        .join(format!("{}{}", hash_mod::to_hex(att_id), FILE_EXT))
}

/// Save a DSSE envelope under `root` for `commit`.
///
/// Returns the `(att_id, full_path)` pair. Idempotent via content-
/// addressing: identical bytes for the same commit produce the same
/// path and skip the write.
///
/// # Errors
/// * [`Error::EnvelopeTooLarge`] if `bytes.len() > MAX_ENVELOPE_SIZE`.
/// * [`Error::Io`] on filesystem failures.
///
/// # Panics
/// Panics if `envelope_path` (an internal helper) returns a path with
/// no parent or no file name. Both invariants are upheld by the path
/// builder, so this only fires on programmer error.
pub fn save(root: &Path, commit: &Hash, bytes: &[u8]) -> Result<(Hash, PathBuf), Error> {
    if bytes.len() > MAX_ENVELOPE_SIZE {
        return Err(Error::EnvelopeTooLarge {
            len: bytes.len(),
            max: MAX_ENVELOPE_SIZE,
        });
    }
    let att_id = env_mod::attestation_id(bytes);
    let final_path = envelope_path(root, commit, &att_id);

    // Idempotency: filename IS the BLAKE3 of the bytes, so presence
    // implies equality. One stat beats a full read+compare.
    if final_path.exists() {
        return Ok((att_id, final_path));
    }

    let parent = final_path.parent().expect("envelope_path has parent");
    fs::create_dir_all(parent).map_err(|e| Error::Io(format!("mkdir: {e}")))?;

    write_atomic(&final_path, bytes)?;
    Ok((att_id, final_path))
}

/// Load an envelope by absolute path (typically returned by [`list`] or
/// [`save`]).
///
/// # Errors
/// * [`Error::EnvelopeTooLarge`] if the file exceeds `MAX_ENVELOPE_SIZE`.
/// * [`Error::Io`] on read failure (incl. not-found).
pub fn load(path: &Path) -> Result<env_mod::Envelope, Error> {
    let meta =
        fs::metadata(path).map_err(|e| Error::Io(format!("stat {}: {e}", path.display())))?;
    // Cast via try_from so 32-bit targets can't truncate the size check.
    let size_usize = usize::try_from(meta.len()).unwrap_or(usize::MAX);
    if size_usize > MAX_ENVELOPE_SIZE {
        return Err(Error::EnvelopeTooLarge {
            len: size_usize,
            max: MAX_ENVELOPE_SIZE,
        });
    }
    let bytes = fs::read(path).map_err(|e| Error::Io(format!("read {}: {e}", path.display())))?;
    env_mod::decode(&bytes)
}

/// List every envelope path attached to `commit`, sorted ascending by
/// att-id (byte-wise on the raw 32-byte hash).
///
/// An unattested commit returns an empty `Vec` — NOT an error, per
/// SPEC-ATTESTATIONS §7.1.
///
/// # Errors
/// * [`Error::Io`] on directory read failure other than not-found.
pub fn list(root: &Path, commit: &Hash) -> Result<Vec<PathBuf>, Error> {
    let commit_dir = root.join(SUBDIR).join(hash_mod::to_hex(commit));
    let read_dir = match fs::read_dir(&commit_dir) {
        Ok(rd) => rd,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Ok(Vec::new()),
        Err(e) => return Err(Error::Io(format!("readdir {}: {e}", commit_dir.display()))),
    };

    // Collect (att_id, path) so we can sort by hash.
    let mut entries: Vec<(Hash, PathBuf)> = Vec::new();
    for entry in read_dir {
        let entry = entry.map_err(|e| Error::Io(format!("readdir entry: {e}")))?;
        if !entry.file_type().is_ok_and(|t| t.is_file()) {
            continue;
        }
        let name = entry.file_name();
        let Some(name_str) = name.to_str() else {
            continue;
        };
        let Some(id) = parse_att_filename(name_str) else {
            continue;
        };
        entries.push((id, entry.path()));
    }
    entries.sort_by_key(|e| e.0);
    Ok(entries.into_iter().map(|(_, p)| p).collect())
}

/// Remove a single attestation. Idempotent: removing a non-existent
/// att-id (or a commit that has no attestations at all) returns
/// successfully. When the commit directory becomes empty, it is
/// removed as well — tolerating `DirNotEmpty` if a concurrent writer
/// landed a new envelope in the interim.
///
/// # Errors
/// * [`Error::Io`] on filesystem failures other than not-found / dir-
///   not-empty during cleanup.
pub fn remove(root: &Path, commit: &Hash, att_id: &Hash) -> Result<(), Error> {
    let final_path = envelope_path(root, commit, att_id);
    match fs::remove_file(&final_path) {
        Ok(()) => {}
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Ok(()),
        Err(e) => return Err(Error::Io(format!("rm {}: {e}", final_path.display()))),
    }
    // Try to GC the now-possibly-empty commit dir.
    let commit_dir = root.join(SUBDIR).join(hash_mod::to_hex(commit));
    match fs::remove_dir(&commit_dir) {
        Ok(()) => {}
        Err(e)
            if matches!(
                e.kind(),
                std::io::ErrorKind::NotFound | std::io::ErrorKind::DirectoryNotEmpty
            ) => {}
        Err(e) => return Err(Error::Io(format!("rmdir {}: {e}", commit_dir.display()))),
    }
    Ok(())
}

// -- internals --

/// Parse `<64-hex-of-att-id>.dsse` → att-id, or return `None`.
fn parse_att_filename(name: &str) -> Option<Hash> {
    let stem = name.strip_suffix(FILE_EXT)?;
    if stem.len() != HEX_LEN {
        return None;
    }
    hash_mod::from_hex(stem).ok()
}

static TEMP_SEQ: AtomicU64 = AtomicU64::new(0);

fn write_atomic(final_path: &Path, bytes: &[u8]) -> Result<(), Error> {
    let parent = final_path.parent().expect("final_path has parent");
    let file_name = final_path
        .file_name()
        .expect("final_path has file name")
        .to_string_lossy();
    let pid = process::id();
    let seq = TEMP_SEQ.fetch_add(1, Ordering::Relaxed);
    let tmp_prefix = format!(".{file_name}.tmp.{pid}.{seq}");

    let mut tmp = NamedTempFile::with_prefix_in(tmp_prefix, parent)
        .map_err(|e| Error::Io(format!("tmpfile: {e}")))?;
    tmp.as_file_mut()
        .write_all(bytes)
        .map_err(|e| Error::Io(format!("write tmp: {e}")))?;
    tmp.as_file_mut()
        .sync_all()
        .map_err(|e| Error::Io(format!("fsync tmp: {e}")))?;
    tmp.persist(final_path)
        .map_err(|e| Error::Io(format!("persist: {}", e.error)))?;
    sync_parent_dir(parent)?;
    Ok(())
}

#[cfg(unix)]
fn sync_parent_dir(parent: &Path) -> Result<(), Error> {
    match File::open(parent) {
        Ok(dir) => dir
            .sync_all()
            .map_err(|e| Error::Io(format!("fsync dir: {e}"))),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(()),
        Err(e) => Err(Error::Io(format!("open dir: {e}"))),
    }
}

#[cfg(not(unix))]
fn sync_parent_dir(_parent: &Path) -> Result<(), Error> {
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::envelope::{Envelope, Sig};

    fn fake_hash(seed: u8) -> Hash {
        let mut h = [0u8; 32];
        for (i, b) in h.iter_mut().enumerate() {
            *b = seed.wrapping_add(u8::try_from(i).unwrap_or(0));
        }
        h
    }

    fn small_envelope(tag: &[u8]) -> (Vec<u8>, Hash) {
        let env = Envelope {
            payload_type: env_mod::PAYLOAD_TYPE_IN_TOTO.into(),
            payload: tag.to_vec(),
            signatures: vec![Sig {
                keyid: "blake3:aa".into(),
                sig: vec![1, 2, 3],
            }],
        };
        let bytes = env.encode().unwrap();
        let id = env_mod::attestation_id(bytes.as_bytes());
        (bytes.into_bytes(), id)
    }

    #[test]
    fn write_read_roundtrip() {
        let dir = tempfile::tempdir().unwrap();
        let commit = fake_hash(0x11);
        let (bytes, expected_id) = small_envelope(b"{\"k\":1}");

        let (att_id, path) = save(dir.path(), &commit, &bytes).unwrap();
        assert_eq!(att_id, expected_id);
        assert!(path.exists());

        let env = load(&path).unwrap();
        assert_eq!(env.payload, b"{\"k\":1}");
    }

    #[test]
    fn write_is_idempotent() {
        let dir = tempfile::tempdir().unwrap();
        let commit = fake_hash(0x22);
        let (bytes, _) = small_envelope(b"hello");

        let (a, _) = save(dir.path(), &commit, &bytes).unwrap();
        let (b, _) = save(dir.path(), &commit, &bytes).unwrap();
        assert_eq!(a, b);

        let listed = list(dir.path(), &commit).unwrap();
        assert_eq!(listed.len(), 1);
    }

    #[test]
    fn list_returns_sorted_ids() {
        let dir = tempfile::tempdir().unwrap();
        let commit = fake_hash(0x33);

        let mut written: Vec<Hash> = Vec::new();
        for i in 0..5 {
            let (bytes, id) = small_envelope(format!("env-{i}").as_bytes());
            let (got, _) = save(dir.path(), &commit, &bytes).unwrap();
            assert_eq!(got, id);
            written.push(id);
        }

        let listed = list(dir.path(), &commit).unwrap();
        assert_eq!(listed.len(), 5);

        // Sorted ascending byte-wise on filename stem.
        let mut last: Option<String> = None;
        for p in &listed {
            let name = p.file_name().unwrap().to_string_lossy().to_string();
            if let Some(prev) = last {
                assert!(prev < name, "{prev} < {name}");
            }
            last = Some(name);
        }
        for w in written {
            let stem = format!("{}{}", hash_mod::to_hex(&w), FILE_EXT);
            assert!(listed.iter().any(|p| p.file_name().unwrap() == &*stem));
        }
    }

    #[test]
    fn list_on_unattested_commit_is_empty() {
        let dir = tempfile::tempdir().unwrap();
        let commit = fake_hash(0x44);
        let listed = list(dir.path(), &commit).unwrap();
        assert!(listed.is_empty());
    }

    #[test]
    fn remove_is_idempotent_on_missing() {
        let dir = tempfile::tempdir().unwrap();
        let commit = fake_hash(0x66);
        let missing = fake_hash(0x77);
        // No attestations dir at all.
        remove(dir.path(), &commit, &missing).unwrap();

        // Commit dir exists but specific id doesn't.
        let (bytes, _) = small_envelope(b"present");
        let _ = save(dir.path(), &commit, &bytes).unwrap();
        remove(dir.path(), &commit, &missing).unwrap();

        let listed = list(dir.path(), &commit).unwrap();
        assert_eq!(listed.len(), 1);
    }

    #[test]
    fn remove_cleans_up_empty_commit_dir() {
        let dir = tempfile::tempdir().unwrap();
        let commit = fake_hash(0x99);
        let (bytes, _) = small_envelope(b"only one");
        let (att_id, _) = save(dir.path(), &commit, &bytes).unwrap();
        remove(dir.path(), &commit, &att_id).unwrap();

        let listed = list(dir.path(), &commit).unwrap();
        assert!(listed.is_empty());

        let commit_dir = dir.path().join(SUBDIR).join(hash_mod::to_hex(&commit));
        assert!(!commit_dir.exists());
    }

    #[test]
    fn list_ignores_non_dsse_and_non_hex_stems() {
        let dir = tempfile::tempdir().unwrap();
        let commit = fake_hash(0xAA);
        let (bytes, good_id) = small_envelope(b"legit");
        let _ = save(dir.path(), &commit, &bytes).unwrap();

        let commit_dir = dir.path().join(SUBDIR).join(hash_mod::to_hex(&commit));
        // Wrong extension.
        fs::write(commit_dir.join("notes.txt"), b"x").unwrap();
        // Right extension, wrong-shaped stem.
        fs::write(commit_dir.join("zzz.dsse"), b"x").unwrap();
        // Right length, non-hex stem.
        let bad = format!("{}{FILE_EXT}", "g".to_string() + &"0".repeat(63));
        fs::write(commit_dir.join(bad), b"x").unwrap();

        let listed = list(dir.path(), &commit).unwrap();
        assert_eq!(listed.len(), 1);
        let want = format!("{}{FILE_EXT}", hash_mod::to_hex(&good_id));
        assert_eq!(listed[0].file_name().unwrap(), want.as_str());
    }
}
